config/application.rb

```diff
@@ -110,8 +114,23 @@ class Application < Rails::Application
 
     config.autoload_paths += config.eager_load_paths
 
-    # config.load_defaults 6.1
-    # Disable defaults as ActiveRecord::Base.belongs_to_required_by_default = true causes MiqRegion.seed to fail validation on belongs_to maintenance zone
+    # FYI, this is where load_defaults is defined as of 7.2:
+    # https://github.com/rails/rails/blob/d437ae311f1b9dc40b442e40eb602e020cec4e49/railties/lib/rails/application/configuration.rb#L92
+    config.load_defaults 7.0
+
+    # TODO: this is the only change we had from defaults in 7.0.  See secure_headers.rb.  It's 0 in defaults.
+    config.action_dispatch.default_headers["X-XSS-Protection"] = "1; mode=block"
+
+    # TODO: Find and fixed any deprecated behavior.  Opt in later.
+    config.active_support.remove_deprecated_time_with_zone_name = false
+    config.active_support.disable_to_s_conversion = false
+
+    # TODO: If disabled, causes cross repo test failures in content, ui-classic and amazon provider
+    config.active_record.partial_inserts = true
+
+    # Disable this setting as it causes MiqRegion.seed to fail validation on belongs_to maintenance zone.
+    # TODO: We should fix this so we don't need to carry this override.
+    config.active_record.belongs_to_required_by_default = false
```

Comment:
Below are the value before and value with rails 7 default

### Overrides:

```
# Log the deprecations so we can fix them.
config.active_support.remove_deprecated_time_with_zone_name = true
config.active_support.disable_to_s_conversion = true

# Causes test failures in content, amazon provider, and ui-classic without it
config.active_record.partial_inserts = false

# X-XSS-Protection is overridden from the rails 7 default back to "1; mode=block"
# Rails 7:
    config.action_dispatch.default_headers = {
              "X-Frame-Options" => "SAMEORIGIN",
              "X-XSS-Protection" => "0",
              "X-Content-Type-Options" => "nosniff",
              "X-Download-Options" => "noopen",
              "X-Permitted-Cross-Domain-Policies" => "none",
              "Referrer-Policy" => "strict-origin-when-cross-origin"
            }
# Previously:
    config.action_dispatch.default_headers = {
              "X-Frame-Options" => "SAMEORIGIN",
              "X-XSS-Protection" => "1; mode=block"
              "X-Content-Type-Options" => "nosniff",
              "X-Download-Options" => "noopen",
              "X-Permitted-Cross-Domain-Policies" => "none",
              "Referrer-Policy" => "strict-origin-when-cross-origin"
            }
```

### Using default value from rails 7

```
config.action_dispatch.return_only_request_media_type_on_content_type = false
config.action_dispatch.cookies_serializer = :json
config.action_view.button_to_generates_button_tag = true
config.action_view.apply_stylesheet_media_default = false
config.active_support.key_generator_hash_digest_class = OpenSSL::Digest::SHA256
config.active_support.cache_format_version = 7.0
config.active_support.use_rfc4122_namespaced_uuids = true
config.active_support.executor_around_test_case = true
config.action_mailer.smtp_timeout = 5
config.active_record.verify_foreign_keys_for_fixtures = true
config.active_record.automatic_scope_inversing = true
config.action_controller.raise_on_open_redirects = true
config.action_controller.wrap_parameters_by_default = true

# Using default value from rails 7.  Was previously: OpenSSL::Digest::SHA1
config.active_support.hash_digest_class = OpenSSL::Digest::SHA256
```

Reply:
@Fryguy @kbrock do any of the "Using default value from rails 7" section above concern you?  They were either nil/false previously or as provided above.  Overrides section above describes the overrides we have in the application.rb so we can try to resolve them in the future.

Reply:
FYI - cross repo is green but I wanted to review the changes I found between current and going to rails 7 defaults.

Reply:
I'll have to go through them, because I don't really understand each of them.

Reply:
FYI, here's the [git blame for 7.0 defaults](https://github.com/rails/rails/blame/v7.0.8.4/railties/lib/rails/application/configuration.rb#L205-L259).  Note, I've already reviewed the overrides seen above.  I think the others we're accepting the rails 7 defaults should either be covered by tests or something we'll find out when others start playing with it.  The OpenSSL changes may not be tested in our tests that well so it's on the list of more concerning.

Reply:
ok, sounds like we need to use the defaults for the  config.action_dispatch.default_headers which disables X-XSS-Protection. 
It's been deprecated and you should XSS auditor tools in older browsers https://www.zdnet.com/article/google-to-remove-chromes-built-in-xss-protection-xss-auditor/

Reply:
config.action_controller.raise_on_open_redirects = true... sounds like a good change to raise if redirect to is called on untrusted URL

Reply:
Funny enough, there was an issue where someone was pointing out how these defaults are not universally documented.  https://www.github.com/rails/rails/issues/50238

Reply:
Is there a way to just spit out the configuration? (Like a rails CLI command or something?)

Reply:
> Is there a way to just spit out the configuration? (Like a rails CLI command or something?)

Wouldn't that be great?  I couldn't find a way.  I was walking the configuration and trying to print it and it was a mess.  A git diff of the configuration would be great.

For now, I think we run the rake task to update our configuration each time we try to do upgrades and check the new defaults.  I think it won't impose new defaults on existing apps so you'd have to bring in new configuration changes and then review the new defaults to determine which you want to accept or reject.

I think we can accept several of these and make plans to fix the rest.  I think the X-XSS-Protection has been proven to be unusable so we can probably just drop that unless we know why we need to keep it.

Reply:
> config.action_controller.raise_on_open_redirects = true... sounds like a good change to raise if redirect to is called on untrusted URL

Fun. Redirecting to untrusted URLs was specifically highlighted in a security course I just took.

Reply:
I'm going to test with x-xss-protection disabled.  It sounds like that is the way going forward for secure headers, rails and even browsers not supporting it.

See: https://www.github.com/github/secure_headers/issues/439